mfe/mfe.go

```diff
@@ -0,0 +1,1209 @@
+		substructuresEnergy += en
+		energyContributions = append(energyContributions, substructureContributions...)
+
+		/* enclosedFivePrimeIdx must have a pairing partner */
+		enclosedThreePrimeIdx := pairTable[enclosedFivePrimeIdx]
+
+		/* get type of the enclosed base pair (enclosedFivePrimeIdx,enclosedThreePrimeIdx) */
+		basePairType := encodedBasePairType(fc, enclosedFivePrimeIdx, enclosedThreePrimeIdx)
+
+		enclosedFivePrimeMismatch := fc.encodedSequence[enclosedFivePrimeIdx-1]
+		enclosedThreePrimeMismatch := fc.encodedSequence[enclosedThreePrimeIdx+1]
+
+		multiLoopEnergy += multiLoopStemEnergy(basePairType, enclosedFivePrimeMismatch,
+			enclosedThreePrimeMismatch, fc.energyParams)
+
+		// seek to the next stem
+		enclosedFivePrimeIdx = enclosedThreePrimeIdx + 1
+
+		for enclosedFivePrimeIdx < closingThreePrimeIdx && pairTable[enclosedFivePrimeIdx] == -1 {
+			enclosedFivePrimeIdx++
+		}
+
+		// add unpaired nucleotides
+		nbUnpairedNucleotides += enclosedFivePrimeIdx - enclosedThreePrimeIdx - 1
+	}
+
+	// add bonus energies for unpaired nucleotides
+	multiLoopEnergy += nbUnpairedNucleotides * fc.energyParams.multiLoopBase
+
+	// Add a energy contribution for this multi-loop
+	contribution := EnergyContribution{
+		closingFivePrimeIdx:  closingFivePrimeIdx,
+		closingThreePrimeIdx: closingThreePrimeIdx,
+		energy:               multiLoopEnergy,
+		loopType:             MultiLoop,
+	}
+	energyContributions = append(energyContributions, contribution)
+	return multiLoopEnergy + substructuresEnergy, energyContributions
+}
+
+/**
+ *  Compute the energy contribution of a multi-loop stem.
+ *
+ *  Given a base pair (i,j) (encoded by `basePairType`), compute the
+ *  energy contribution including dangling-end/terminal-mismatch contributions.
+ *
+ *  @param  basePairType          The encoded base pair type of (i, j) (see `basePairType()`)
+ *  @param  fivePrimeMismatch     The encoded nucleotide directly adjacent (in the 5' direction) to i (may be -1 if index of i is 0)
+ *  @param  threePrimeMismatch    The encoded nucleotide directly adjacent (in the 3' direction) to j (may be -1 if index of j is len(sequence) - 1)
+ *  @param  energyParams          The pre-computed energy parameters
+ *  @return                       The energy contribution of the introduced multi-loop stem
+ */
+func multiLoopStemEnergy(basePairType, fivePrimeMismatch, threePrimeMismatch int, energyParams *energyParams) int {
+	var energy int = energyParams.mismatchMultiLoop[basePairType][fivePrimeMismatch][threePrimeMismatch] +
+		energyParams.multiLoopIntern[basePairType]
+
+	if basePairType > 2 {
+		// The closing base pair is not a GC or CG pair (could be GU, UG, AU, UA, or non-standard)
+		energy += energyParams.terminalAU
+	}
+
+	return energy
+}
+
+// Returns the minimum of two ints
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func scaleEnergyParams(temperature float64) *energyParams {
```

Comment:
This function's control flow is absolute chaos and we need a ton of comments to explain what's going on here.

Reply:
Do we? We're just setting the values for the energy parameters so we're just looping through the matrices and setting values. Not sure if the control flow needs more information.